File: android/app/src/main/kotlin/run/nuri/getagrip/store/DayClock.kt
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

package run.nuri.getagrip.store

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import androidx.compose.runtime.Stable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.runtime.ProvidableCompositionLocal
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.core.content.ContextCompat
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import run.nuri.getagrip.engine.DayStamp
import java.time.Instant

/// The app's single source of "today".
///
/// `DayStamp.today()` read in a composable only updates when Compose happens to recompose
/// it, so an app open across midnight kept showing yesterday ("1 of 2 today", yesterday's
/// strip cell). The day is observable state of its own, and every surface recomputes
/// together.
///
/// TRANSLATION NOTE (Sources/Store/DayClock.swift): UIKit's one `significantTimeChange`
/// becomes three Android broadcasts — `ACTION_DATE_CHANGED`, `ACTION_TIME_CHANGED`,
/// `ACTION_TIMEZONE_CHANGED` — all landing in `refresh()`. They are protected implicit
/// broadcasts, so registered at RUNTIME, not in the manifest.
@Stable
class DayClock(today: DayStamp = DayStamp.today()) {

    /// `today` is a test seam: crossing midnight is the most expensive behaviour to verify
    /// by waiting.
    var today: DayStamp by mutableStateOf(today)
        private set

    /// Also called on resume: a device asleep across midnight may not deliver the broadcast
    /// until foregrounded.
    fun refresh() {
        val now = DayStamp.today()
        if (now != today) today = now
    }

    /// Tests only. Kept off `refresh`'s path so no shipping code can pin the day against
    /// the system clock.
    fun advance(to: DayStamp) {
        if (to != today) today = to
    }

    /// The training day turns at `DayStamp.ROLLOVER_HOUR` (04:00), which no broadcast
    /// marks. So the clock sleeps until the next rollover, refreshes, tells the store, and
    /// repeats for the process's life; a dead process is caught by the resume refresh. iOS:
    /// `DayClock.armRolloverRefresh`.
    fun scheduleRolloverRefresh(scope: CoroutineScope, onDayMayHaveChanged: () -> Unit): Job =
        scope.launch {
            while (isActive) {
                val wait = DayStamp.nextRollover(Instant.now()).toEpochMilli() - System.currentTimeMillis()
                delay(maxOf(1_000L, wait))
                refresh()
                onDayMayHaveChanged()
            }
        }
}

/// Registered by the Application for the life of the process. **Refreshes the clock, THEN
/// asks the store to recompute** — the store only REACTS to the clock (see
/// `TemplateStore.refreshIfDayChanged`). `DoigtApp`'s `scenePhase == .active` order.
class DayClockReceiver(
    private val clock: DayClock,
    private val onDayMayHaveChanged: () -> Unit,
) : BroadcastReceiver() {

    override fun onReceive(context: Context?, intent: Intent?) {
        clock.refresh()
        onDayMayHaveChanged()
    }

    /// `ContextCompat`: `RECEIVER_NOT_EXPORTED` only means what it says from API 33 (minSdk
    /// is 31). Not exported regardless — protected system broadcasts only.
    fun register(context: Context) {
        val filter = IntentFilter().apply {
            addAction(Intent.ACTION_DATE_CHANGED)
            addAction(Intent.ACTION_TIME_CHANGED)
            addAction(Intent.ACTION_TIMEZONE_CHANGED)
        }
        ContextCompat.registerReceiver(
            context.applicationContext, this, filter, ContextCompat.RECEIVER_NOT_EXPORTED,
        )
    }
}

/// See `LocalDeviceStore` for why this is `staticCompositionLocalOf`.
val LocalDayClock: ProvidableCompositionLocal<DayClock> = staticCompositionLocalOf {
    error("LocalDayClock was read outside a CompositionLocalProvider")
}

File: Sources/Store/DayClock.swift
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

import Foundation
import Observation
#if !os(watchOS)
import UIKit
#endif

/// The app's single source of "today".
///
/// Reading `DayStamp.today()` from a view body only re-evaluates when SwiftUI happens to
/// invalidate that view, so an app left open across midnight kept rendering yesterday.
/// The day gets an observable of its own, and every surface recomputes together.
///
/// `significantTimeChangeNotification` fires at local midnight, on a manual clock change
/// and on a time-zone change. The TRAINING day turns at `DayStamp.rolloverHour`, which
/// nothing announces, so the clock books its own wake-up (`armRolloverRefresh`).
@Observable @MainActor
final class DayClock {
    private(set) var today: DayStamp
    /// The store reacts after the clock has changed, independent of notification order.
    @ObservationIgnored var onDayChanged: (@MainActor () -> Void)?
    /// The pending wake-up for the next rollover. Holds `self` weakly, so a clock that is
    /// gone simply lets it lapse; re-armed by every refresh.
    @ObservationIgnored private var rolloverRefresh: Task<Void, Never>?

    /// Tokens live in a box that unregisters itself: Swift 6 forbids a nonisolated
    /// `deinit` from touching a non-Sendable stored property of a `@MainActor` class,
    /// and the box's own deinit has no such restriction.
    private final class ObserverTokens: @unchecked Sendable {
        var tokens: [any NSObjectProtocol] = []
        deinit {
            let center = NotificationCenter.default
            for token in tokens { center.removeObserver(token) }
        }
    }

    @ObservationIgnored private let observers = ObserverTokens()

    /// The `today:` parameter is a test seam: crossing midnight is too expensive to
    /// verify by waiting.
    init(today: DayStamp = .today()) {
        self.today = today
        let center = NotificationCenter.default
        #if os(watchOS)
        // No application object on the wrist. Foundation's day-change post is the same
        // midnight; the manual-clock-change case UIKit also folds in is covered by the
        // foreground refresh.
        let names: [Notification.Name] = [.NSCalendarDayChanged, .NSSystemTimeZoneDidChange]
        #else
        let names: [Notification.Name] = [
            UIApplication.significantTimeChangeNotification,
            .NSSystemTimeZoneDidChange,
        ]
        #endif
        observers.tokens = names.map { name in
            center.addObserver(forName: name, object: nil, queue: .main) { [weak self] _ in
                Task { @MainActor in self?.refresh() }
            }
        }
        armRolloverRefresh()
    }

    /// Also called on foreground: a device that was asleep across midnight — or across
    /// the 04:00 rollover — may not deliver the time-change notification, and cannot run
    /// the rollover task, until the app is active again.
    func refresh() {
        let now = DayStamp.today()
        advance(to: now)
        armRolloverRefresh()
    }

    /// Sleeps until the next `DayStamp.rolloverHour` and refreshes itself; a process
    /// suspended across it is caught by the foreground refresh instead. Continuous
    /// clock, so a phone that dozed with the app foregrounded still wakes on time.
    private func armRolloverRefresh() {
        rolloverRefresh?.cancel()
        let interval = DayStamp.nextRollover(after: .now).timeIntervalSinceNow
        rolloverRefresh = Task { [weak self] in
            try? await Task.sleep(for: .seconds(max(1, interval)), clock: .continuous)
            guard !Task.isCancelled else { return }
            self?.refresh()
        }
    }

    /// Tests only in practice — nothing in the app calls it. Kept out of `refresh`'s
    /// path deliberately, so no shipping code can pin the day to a value the system
    /// clock disagrees with.
    func advance(to day: DayStamp) {
        guard day != today else { return }
        today = day
        onDayChanged?()
    }
}
